feat: Split serverless plugins into their own pages (#6047)
Kong/docs.konghq.com: app/_hub/kong-inc/post-function/how-to/_index.md

---
title: Getting started with the Post-function plugin
nav_title: Getting started with Post-function
---

> Let's test out the post-function plugin by changing some request header names.

Reviewer: ```suggestion
Let's test out the Post-function plugin by changing some request header names.
```

We'll try it out on two types of headers: headers set via a plugin (in this case, 
[Rate Limiting](/hub/kong-inc/rate-limiting/)), and latency headers from {{site.base_gateway}}).

* The Rate Limiting plugin returns headers such as `X-RateLimit-Remaining-<time>` and `X-RateLimit-Limit-<time>`, 
where `<time>` is the configured time span for the limit.
* {{site.base_gateway}} adds latency headers to responses, such as `X-Kong-Upstream-Latency` and `X-Kong-Proxy-Latency`.

> While you can turn these headers on or off in `kong.conf`, they have fixed names that can't be changed via direct configuration. 

Reviewer: ```suggestion
While you can turn these headers on or off in `kong.conf`, they have fixed names that can't be configured. 
```

To change any header names, you have to set up a Post-function plugin instance that runs in the `header_filter` phase.

> 1. Create a service in {{site.base_gateway}}:

Reviewer: ```suggestion
1. Create a {{site.base_gateway}} service:
```

      kong_rl_headers["x-ratelimit-limit-day"]="X-Rlld"
      kong_rl_headers["x-ratelimit-remaining-day"]="X-Rlrd"
      kong_rl_headers["x-ratelimit-limit-month"]="X-Rlln"
      kong_rl_headers["x-ratelimit-remaining-month"]="X-Rlrn"
      kong_rl_headers["x-ratelimit-limit-year"]="X-Rlly"
      kong_rl_headers["x-ratelimit-remaining-year"]="X-Rlry"

      local headers = kong.response.get_headers()
      for k, v in pairs(headers) do
        if kong_rl_headers[k] ~= nil then
          kong.response.set_header(kong_rl_headers[k], v)
          kong.response.clear_header(k)
        end
      end

      -- Add custom headers for latency
      kong.response.set_header("My-Custom-Proxy-Latency", ngx.ctx.KONG_PROXY_LATENCY)
      kong.response.set_header("My-Custom-Upstream-Latency", ngx.ctx.KONG_WAITING_TIME)

    end
    ```

1. Add the serverless function to the appropriate Kong entity by enabling the post-function plugin. 
For example, let's apply it to the `test` route:

    ```sh
    curl -i -X POST https://localhost:8001/routes/test/plugins \
      --form "name=post-function" \
      --form "config.header_filter=@/tmp/serverless.lua"
    ```

> 1. Verify the changes. Make a call to the Admin API and check the response header names have been changed:

Reviewer: ```suggestion
1. Verify the changes. Make a call to the Admin API and check that the response header names have been changed:
```

    ```sh
    curl -v http://localhost:8000/test
    ```

    The response should show the new header names:

    ```
    *   Trying 127.0.0.1)...
    * TCP_NODELAY set
    * Connected to kong (127.0.0.1)) port 8000 (#0)
    > GET /test HTTP/1.1
    > Host: localhost:8000
    > User-Agent: curl/7.86.0
    > Accept: */*
    >
    < HTTP/1.1 200 OK
    < Content-Type: application/json
    < Content-Length: 374
    < Connection: keep-alive
    < Server: gunicorn/19.9.0
    < Date: Mon, 27 Apr 2023 09:27:20 GMT
    < Access-Control-Allow-Origin: *
    < Access-Control-Allow-Credentials: true
    < X-Rlls: 2
    < X-Rllm: 5
    < X-Rlrm: 3
    < X-Rlrs: 1
    < My-Custom-Proxy-Latency: 1
    < My-Custom-Upstream-Latency: 4
    ```

This is just a small demonstration of the power this plugin grants. You were
able to dynamically inject Lua code into the header phase to dynamically
transform the request without creating a custom plugin or

> reloading/redeploying Kong.

Reviewer: ```suggestion
reloading and redeploying {{site.base_gateway}}.
```
